File: 4.graph/8.minimum-island/minimum-island.go
// Write a function, minimumIsland, that takes in a grid containing Ws and Ls. W represents water and L represents land. The function should return the size of the smallest island. An island is a vertically or horizontally connected region of land.

// You may assume that the grid contains at least one island.

package main

import (
	"fmt"
	"math"
)

type Graph struct {
	Nodes map[string][]string
}

func islandCount(grid [][]rune) int {
	minIsland := math.MaxInt
	visited := make(map[string]bool)

	for i := 0; i < len(grid); i++ {
		for j := 0; j < len(grid[0]); j++ {
			if grid[i][j] == 'W' {
				continue
			}
			if visited[fmt.Sprintf("%d,%d", i, j)] {
				continue
			}
			islandCount := exploreIsland(grid, i, j, visited)
			if islandCount < minIsland {
				minIsland = islandCount
			}
		}
	}
	return minIsland
}

func exploreIsland(grid [][]rune, i, j int, visited map[string]bool) int {
	rowInbounds := 0 <= i && i < len(grid)
	colInbounds := 0 <= j && j < len(grid[0])
	if !rowInbounds || !colInbounds {
		return 0
	}
	if visited[fmt.Sprintf("%d,%d", i, j)] {
		return 0
	}
	if grid[i][j] == 'W' {
		return 0
	}
	if grid[i][j] == 'L' {
		visited[fmt.Sprintf("%d,%d", i, j)] = true
	}
	return 1 +
		exploreIsland(grid, i+1, j, visited) +
		exploreIsland(grid, i-1, j, visited) +
		exploreIsland(grid, i, j+1, visited) +
		exploreIsland(grid, i, j-1, visited)
}

func main() {
	grid := [][]rune{
		{'W', 'L', 'W', 'W', 'W'},
		{'W', 'L', 'W', 'W', 'W'},
		{'W', 'W', 'W', 'L', 'W'},
		{'W', 'W', 'L', 'L', 'W'},
		{'L', 'W', 'W', 'L', 'L'},
		{'L', 'L', 'W', 'W', 'W'},
	}

	fmt.Println(islandCount(grid))
}

File: 4.graph/8.minimum-island/minimum-island.js
// Write a function, minimumIsland, that takes in a grid containing Ws and Ls. W represents water and L represents land. The function should return the size of the smallest island. An island is a vertically or horizontally connected region of land.

// You may assume that the grid contains at least one island.

const minimumIsland = grid => {
    let minIsland = Infinity;
    const visited = new Set();

    for (let i = 0; i < grid.length; i++) {
        for (let j = 0; j < grid[0].length; j++) {
            if (grid[i][j] === 'W') continue;
            if (visited.has(`${i},${j}`)) continue;
            const isLandCount = exploreIsland(grid, i, j, visited);
            if (isLandCount < minIsland) minIsland = isLandCount;
        }
    }
    return minIsland;
};

const exploreIsland = (grid, i, j, visited) => {
    const rowInbounds = 0 <= i && i < grid.length;
    const colInbounds = 0 <= j && j < grid[0].length;
    if (!rowInbounds || !colInbounds) return 0;
    if (visited.has(`${i},${j}`)) return 0;
    if (grid[i][j] === 'W') return 0;
    if (grid[i][j] === 'L') visited.add(`${i},${j}`);
    const right = exploreIsland(grid, i + 1, j, visited);
    const left = exploreIsland(grid, i - 1, j, visited);
    const down = exploreIsland(grid, i, j + 1, visited);
    const up = exploreIsland(grid, i, j - 1, visited);
    return 1 + right + left + down + up;
};

const grid = [
    ['W', 'L', 'W', 'W', 'W'],
    ['W', 'L', 'W', 'W', 'W'],
    ['W', 'W', 'W', 'L', 'W'],
    ['W', 'W', 'L', 'L', 'W'],
    ['L', 'W', 'W', 'L', 'L'],
    ['L', 'L', 'W', 'W', 'W'],
];

console.log(minimumIsland(grid)); // -> 2
